docs/05-actor-model/idea.md

```diff
@@ -0,0 +1,327 @@
+being chiefs dispatcher which would collect all orders from
+sellers, and balance them across chiefs for some reason.
+
+## The action
+
+Actors are the model entities, but to properly communicate with
+them, we need some kind of protocol to communicate with them.
+Every actor is capable of performing several actions. In my
+previous KFC example the only action seller can do is "Charge
+payment and create order". However it is not always a case -
+our chefs was proficient with performing both "Prepare fries"
+and "Prepare Grander" actions - and also many more.
+
+So when we want to do something in an actor system, we schedule
+some action to the actor being the closes to us, very often with
+some additional parameter (as we can pick if we want to exchange
+fries with salad).
+
+However naming the action after the exact thing which happen
+in the very contract would be misleading. Take a look at KFC
+example once again. As I mentioned, the action performed by seller
+is "Charge payment and create order". The problem is, that for
+the client which schedules this action, it doesn't matter what
+exactly is the responsibility of actor himself - what client is
+actually scheduling is "Prepare Meal" with some description
+of what exactly to prepare. So we can say, that the action is
+the thing performed by contract itself, plus all the sub-actions it
+schedules.
+
+## Multi-stage actions
+
+So as the whole idea makes some sense, there is the biggest
+problem created by actor model: what if I want to perform some
+action in my contract, but to completely finalize some steps,
+the contract have to make sure that some sub-action he
+scheduled are actually finished?
+
+Imagine, that in the previous KFC situation there is no dedicated
+Waiter, instead Seller is serving you meal when chefs finished
+their job.
+
+This kind of pattern is so important and common, that in CosmWasm
+we developed special way to handle it, which is dedicated `Reply`
+action.
+
+So when Seller is scheduling actions for chiefs, he assigns some
+number to this action (like order id), and passes it to chiefs.
+He also remembers how much actions he scheduled to every order id.
+Now every time chief is finished with his action, he would call the
+special `Reply` action on Seller, in which he would pass back the
+order id. Then Seller would decrease number of actions left for
+this order, and if it reached zero, he would serve meal.
+
+Now you can say, that the `Reply` action is completely not needed,
+as Chiefs could just schedule any arbitrary action on Seller, like
+`Serve`, why there is the special `Reply` for? The reason is
+abstraction and reusability. The Chiefs task is to prepare a meal,
+and that is all. There is no reason for him to know why he is
+even preparing Fries - if it is part of the bigger task (like order
+for client), or seller is just hungry. There is actually possible,
+that no only seller is eligible to call chief for food - possibly
```

Comment:
that not only Seller is eligible to call chief for food - possibly any restaurant employee can do that just for themselves.